content/300-guides/150-testing/100-unit-testing.mdx

```diff
@@ -0,0 +1,311 @@
+
+Prisma uses TypeScript and Jest can be setup to use TypeScript with the help of the [`ts-jest`](https://github.com/kulshekhar/ts-jest) package.
+
+```terminal
+npm install --save-dev jest typescript ts-jest @types/jest
+```
+
+Next create a `jest.config.js` file at your projects root, adding the following configuration:
+
+```js file=jest.config.js
+module.exports = {
+  clearMocks: true,
+  preset: 'ts-jest',
+  testEnvironment: 'node',
+  setupFilesAfterEnv: ['<rootDir>/singleton.ts'],
+}
+```
+
+The above configuration declares a path to a `singleton.ts` file, this can be called anything you like, the following guide will name it `singleton.ts`. The module the path points towards is the code that runs setting up the testing framework before each test file is executed.
+
+Add a script to your `package.json` file which will run the tests. The `-i` flag will ensure that the tests are run one at a time to avoid any race conditions. The scripts will look for a folder at your projects root called `__tests__` (double\_\_underscore), and run any tests within files, within that folder.
+
+```json file=package.json
+  "scripts": {
+    ...otherScripts
+    "test": "jest -i"
+  },
+```
+
+</details>
+
+## Mocking the Prisma client
+
+To ensure your unit tests are isolated from external factors you can mock the Prisma client, this means you get the benefits of being able to use your schema (**_type-safety_**), without having to make actual calls to your database when your tests are run.
+
+This guide will cover two approaches to mocking the client, a singleton instance and dependency injection. Both have their merits depending on your use cases. To help with mocking the client the [`jest-mock-extended`](https://github.com/marchaos/jest-mock-extended) package will be used.
+
+```terminal
+npm install jest-mock-extended --save-dev
+```
+
+### Singleton
+
+1. Create a file at your projects root called `client.ts` and add the following code. This will instantiate a Prisma client instance.
+
+```ts file=client.ts
+import { PrismaClient } from '@prisma/client'
+
+const prisma = new PrismaClient()
+export default prisma
+```
+
+2. Next create a file called `singleton.ts` at your projects root and add the following:
+
+```ts file=singleton.ts
+import { PrismaClient } from '@prisma/client'
+import { mockDeep, mockReset, MockProxy } from 'jest-mock-extended'
+
+import prisma from './client'
+
+jest.mock('./client', () => ({
+  __esModule: true,
+  default: mockDeep<PrismaClient>(),
+}))
+
+beforeEach(() => {
+  mockReset(prismaMock)
+})
+
+export const prismaMock = (prisma as unknown) as MockProxy<PrismaClient>
+```
+
+The singleton file tells Jest to mock a default export (the Prisma client instantiated in `./client.ts`), and uses the `mockDeep` method from `jest-mock-extended` to enable access to the objects and methods available on the Prisma client. It then resets the mocked instance before each test is run.
+
+### Dependency injection
+
```

Comment:
Introduce - "To mock the client with dependency injection"

Also needs steps so that it matches the singleton examples